File: 2262-solving-questions-with-brainpower/solving-questions-with-brainpower.java
class Solution {
    public long mostPoints(int[][] questions) {
        int n = questions.length;
        long[] dp = new long[n + 1];
        dp[n - 1] = questions[n - 1][0];

        for (int i = n - 2; i >= 0; i--) {
            dp[i] = questions[i][0];

            int nextIndex = questions[i][1] + i + 1;
            long solve = questions[i][0] + (nextIndex < n ? dp[nextIndex] : 0);
            long skip = dp[i + 1];

            dp[i] = Math.max(solve, skip);
        }

        return dp[0];
    }
}

File: 2262-solving-questions-with-brainpower/solving-questions-with-brainpower.go
func mostPoints(questions [][]int) int64 {
    n := len(questions);
    dp := make([]int64, n + 1);
    dp[n - 1] = int64(questions[n - 1][0]);
    
    for i := n - 2; i >= 0; i-- {
        dp[i] = int64(questions[i][0]);

        nextIndex := i + questions[i][1] + 1;
        solve := int64(questions[i][0]);
        if nextIndex < n {
            solve += dp[nextIndex];
        }

        dp[i] = max(solve, dp[i + 1]);
    }

    return dp[0];
}

func max(a, b int64) int64 {
    if a > b {
        return a;
    }

    return b;
}